pkg/engine/constraints/application_constraint.go

```diff
@@ -0,0 +1,72 @@
+package constraints
+
+import (
+	"errors"
+
+	"github.com/klothoplatform/klotho/pkg/core"
+)
+
+type (
+	// ApplicationConstraint is a struct that represents constraints that can be applied on the entire resource graph
```

Comment:
im going to do a large documentation pass towards the end of this if thats ok, but can toss in some samples quick for now.

Just avoiding doing too much in case it changes